### ISSUE TITLE
GPU/CUDA support for PBS and gate evaluation

### ISSUE THREAD

**Issue by jimouris** [zama-ai/tfhe-rs#237]
**What is the problem you want to solve and can not with the current version?**
Hi, I was wondering if you have considered GPU/CUDA acceleration for gate evaluation and bootstrapping.

**Describe the solution you'd like**
Ideally, adding CUDA/GPU support :) I'd be interested in looking into it as well.

**Describe alternatives you've considered**
I saw that `concrete` has some CUDA support but I was mostly interested in `tfhe-rs` rather than `concrete`.

**Comment by jborfila** [comment]
Hi @jimouris,
At the moment the GPU is not supported , but we might have a look at it at some point.
This issue stays open, and let you know about any progress in this direction :)

**Comment by jopasserat** [comment]
Following up on this: can you comment even briefly on the design choice that make it more appealing to support GPUs at the concrete level (as announced [here](https://community.zama.ai/t/concrete-core-v1-0-0-gamma-with-gpu-acceleration/234)) rather than in tfhe-rs directly? :thinking:

**Comment by IceTDrinker** [comment]
hello @jopasserat the announcement you link is a bit old and does not reflect the current state of our products.

Concrete is now Zama's FHE Compiler see https://www.zama.ai/post/zama-concrete-fully-homomorphic-encryption-compiler

As a matter of fact the older brother of TFHE-rs, known at the time as concrete-core, had a dedicated GPU backend, I don't see why it would be any different to bring GPU support to the higher level APIs of TFHE-rs (which include the integer API you knew from the old concrete, before the product scope change to a compiler).

So in short:
If TFHE-rs gets GPU support it would most likely include GPU primitives itself for the lower level APIs and that would then be re-exposed in the higher level layers to allow running e.g. large integer arithmetic on GPU

Cheers

**Comment by jopasserat** [comment]
thanks @IceTDrinker yeah sorry wrong link I was aware of the renaming. I guess my question is more about the link, if any, between tfhe-rs and new concrete. I was under the impression that tfhe-rs was the core implementation and concrete was the compiler + optimiser + domain specific APIs (ML, ...).  
But now I see that concrete has its own CPU and CUDA backends and doesn't seem to rely on tfhe-rs :thinking:

**Comment by IceTDrinker** [comment]
@jopasserat yes you are correct, at the moment concrete relies on an implementation derived from concrete-core (called concrete-cpu), at some point to ease integration and for dependency management reasons it made sense. Now we are looking to streamline our integration so the state of things should change down the line

**Comment by agnesLeroy** [comment]
Hey @jopasserat! You may wish to check TFHE-rs version 0.5.1, as it offers GPU support: https://docs.zama.ai/tfhe-rs/how-to/run_on_gpu.

Best regards,
Agnes
